x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/extractor/fields/ExtractedField.java

```diff
@@ -88,12 +99,86 @@ public static ExtractedField newField(String alias, String name, ExtractionMetho
                 List<Object> values = keyValue.getValues();
-                return values.toArray(new Object[values.size()]);
+                return values.toArray(new Object[0]);
             }
             return new Object[0];
         }
     }
 
+    private static class GeoField extends ExtractedField {
+
+        private final ExtractedField internalExtractor;
+        private final WellKnownText wkt = new WellKnownText();
+
+        GeoField(String alias, String name, ExtractionMethod extractionMethod) {
+            super(alias, name, extractionMethod);
+            internalExtractor = extractionMethod.equals(ExtractionMethod.SOURCE) ?
+                new FromSource(alias, name, extractionMethod) :
+                new FromFields(alias, name, extractionMethod);
+        }
+
+        @Override
+        public Object[] value(SearchHit hit) {
+            Object[] value = internalExtractor.value(hit);
+            if (value.length == 2 && internalExtractor.getExtractionMethod().equals(ExtractionMethod.SOURCE)) { // geo_point as array
+                return new Object[] {value[0] + "," + value[1]};
+            }
+            if (value.length != 1) {
+                throw new IllegalStateException("Unexpected value count for a geo point field: " + value);
+            }
+            if (value[0] instanceof String) {
+                value[0] = handleString((String) value[0]);
+            } else if(value[0] instanceof Map<?, ?>) {
+                @SuppressWarnings("unchecked")
+                Map<String, Object> geoObject = (Map<String, Object>) value[0];
+                value[0] = handleObject(geoObject);
+            } else if(value[0] instanceof List<?>) {
+                @SuppressWarnings("unchecked")
+                List<Double> coordinates = (List<Double>) value[0];
+                assert coordinates.size() == 2;
+                value[0] = coordinates.get(0) + "," + coordinates.get(1);
+            } else {
+                throw new IllegalStateException("Unexpected value type for a geo point field: " + value[0].getClass());
+            }
+            return value;
+        }
+
+        private String handleString(String geoString) {
+            try {
+                if (geoString.startsWith("POINT")) { // Entry is of the form "POINT (-77.03653 38.897676)"
+                    Geometry geometry = wkt.fromWKT(geoString);
+                    if (geometry.type() != ShapeType.POINT) {
+                        throw new IllegalArgumentException("Unexpected non-point geo type: " + geometry.type().name());
+                    }
+                    Point pt = ((Point)geometry);
+                    return pt.getLat() + "," + pt.getLon();
+                } else if (geoString.contains(",")) { // Entry is of the form "38.897676, -77.03653"
+                    return geoString.replace(" ", "");
+                } else { // This may be a geohash, attempt to decode
+                    Point pt = Geohash.toPoint(geoString);
```

Comment:
I don't think we need to handle this. If I understand correctly, we want to support `get_point` and `geo_shape` here.

`geo_shape` can only be retrieved from `_source` and we basically deal with it in `handleObject`.

For `geo_point` we can force the extraction method to be from doc values. Then, we now the value will be `["<lat>,<long>"]` even if it was indexes as a geohash. Does this make sense?

Reply:
Yeah for sure, do we have guarantees for the doc values? What if they have `doc_values` turned off for the mapped field?

Reply:
I think we should error if we detect a `geo_point` that doesn't have doc values. I don't see the point of mapping a field as `geo_point` only to disable doc values for it.